### PR TITLE
Add new resource and data source for Virtual Hub

### PR DESCRIPTION
**Pull request by neil-yechenwei**

### CODE REVIEW

File: azurerm/resource_arm_virtual_hub_test.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+				),
+			},
+			{
+				Config:      testAccAzureRMVirtualHub_requiresImport(ri, testLocation()),
+				ExpectError: testRequiresImportError("azurerm_virtual_hub"),
+			},
+		},
+	})
+}
+
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub_test.go:86]: Can we get an Update and a Complete test in here too?

**Review by WodansSon — CHANGES_REQUESTED** [review]
@neil-yechenwei The is looking very good, there are only two things I would to see updated. Can we get a couple of more test cases (comment below) and the TOC seems to be out of sync... can you pull from master and get that in sync again?

File: azurerm/resource_arm_virtual_hub_test.go (hunk 1 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub_test.go:86]: Sure. I've added update test and complete test. Thanks.

**Comment by neil-yechenwei** [comment]
> @neil-yechenwei The is looking very good, there are only two things I would to see updated. Can we get a couple of more test cases (comment below) and the TOC seems to be out of sync... can you pull from master and get that in sync again?

I've also updated TOC. Thanks.

File: azurerm/resource_arm_virtual_hub.go [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+				ValidateFunc: validate.NoEmptyStrings,
+			},
+
+			"location": azure.SchemaLocation(),
+
+			"resource_group_name": azure.SchemaResourceGroupNameDiffSuppress(),
+
+			"address_prefix": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: Can we add some validation here? WDYT?

```suggestion
				Optional: true,
				ValidateFunc: validate.NoEmptyStrings,
```

File: azurerm/resource_arm_virtual_hub.go [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+			},
+
+			"express_route_gateway": {
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: can we flatten this attribute to `express_route_gateway_id`?

File: azurerm/resource_arm_virtual_hub.go [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"p2svpn_gateway": {
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: same here, can we flatten this to `p2svpn_gateway_id`?

File: azurerm/resource_arm_virtual_hub.go [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+										Optional: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"tags": tags.Schema(),
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: Can we move this to the bottom of the schema definition?

File: azurerm/resource_arm_virtual_hub.go [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"remote_virtual_network": {
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: can we flatten this attribute to `remote_virtual_network_id`

File: azurerm/resource_arm_virtual_hub.go [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"virtual_wan": {
+				Type:     schema.TypeList,
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: same here, can we flatten this the `virtual_wan_id`?

File: azurerm/resource_arm_virtual_hub.go [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"vpn_gateway": {
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub.go]: and here, can we flatten this to `vpn_gateway_id`?

File: azurerm/data_source_virtual_hub.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+			"location": azure.SchemaLocationForDataSource(),
+
+			"resource_group_name": azure.SchemaResourceGroupNameForDataSource(),
+
+			"address_prefix": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"express_route_gateway": {
```
Comment (WodansSon) [inline azurerm/data_source_virtual_hub.go]: can we flatten this to `express_route_gateway_id`?

File: azurerm/data_source_virtual_hub.go [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"p2svpn_gateway": {
```
Comment (WodansSon) [inline azurerm/data_source_virtual_hub.go]: can we flatten this to `p2svpn_gateway_id`?

File: azurerm/data_source_virtual_hub.go [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+						},
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"remote_virtual_network": {
```
Comment (WodansSon) [inline azurerm/data_source_virtual_hub.go]: can we flatten this to `remote_virtual_network_id`?

File: azurerm/data_source_virtual_hub.go [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+										Computed: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"virtual_wan": {
```
Comment (WodansSon) [inline azurerm/data_source_virtual_hub.go]: same here, can we flatten this to `virtual_wan_id`?

File: azurerm/data_source_virtual_hub.go [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"vpn_gateway": {
```
Comment (WodansSon) [inline azurerm/data_source_virtual_hub.go]: again here, can we flatten this to `vpn_gateway_id`?

File: azurerm/resource_arm_virtual_hub_test.go [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+					resource.TestCheckResourceAttr(resourceName, "route_table.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "route_table.0.routes.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "route_table.0.routes.0.address_prefixes.#", "2"),
+					resource.TestCheckResourceAttr(resourceName, "route_table.0.routes.0.address_prefixes.0", "10.0.2.0/24"),
+					resource.TestCheckResourceAttr(resourceName, "route_table.0.routes.0.address_prefixes.1", "10.0.3.0/24"),
+					resource.TestCheckResourceAttr(resourceName, "route_table.0.routes.0.next_hop_ip_address", "10.0.4.6"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.env", "test"),
+				),
+			},
```
Comment (WodansSon) [inline azurerm/resource_arm_virtual_hub_test.go:132]: can we add a new step where it updates back to the basic state and then have an import check?

**Review by WodansSon — CHANGES_REQUESTED** [review]
Hey @neil-yechenwei , thanks for the PR... this mostly LGTM, I have left a few mostly minor comments in the PR. 🚀

File: azurerm/resource_arm_virtual_hub.go (hunk 2 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: updated

File: azurerm/resource_arm_virtual_hub.go (hunk 5 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: Updated

File: azurerm/resource_arm_virtual_hub.go (hunk 3 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: Updated

File: azurerm/resource_arm_virtual_hub.go (hunk 4 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: Updated

File: azurerm/resource_arm_virtual_hub.go (hunk 6 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: Updated

File: azurerm/resource_arm_virtual_hub.go (hunk 7 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: Updated

File: azurerm/resource_arm_virtual_hub.go (hunk 8 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub.go]: Updated

File: azurerm/data_source_virtual_hub.go (hunk 9 above)
Comment (neil-yechenwei) [reply azurerm/data_source_virtual_hub.go]: Updated

File: azurerm/data_source_virtual_hub.go (hunk 10 above)
Comment (neil-yechenwei) [reply azurerm/data_source_virtual_hub.go]: Updated

File: azurerm/data_source_virtual_hub.go (hunk 11 above)
Comment (neil-yechenwei) [reply azurerm/data_source_virtual_hub.go]: Updated

File: azurerm/data_source_virtual_hub.go (hunk 12 above)
Comment (neil-yechenwei) [reply azurerm/data_source_virtual_hub.go]: Updated

File: azurerm/data_source_virtual_hub.go (hunk 13 above)
Comment (neil-yechenwei) [reply azurerm/data_source_virtual_hub.go]: Updated

File: azurerm/resource_arm_virtual_hub_test.go (hunk 14 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_virtual_hub_test.go:132]: Updated

**Comment by neil-yechenwei** [comment]
@jeffreyCline , Thanks for your comments. I've updated code. Please help to take a look. Thanks.

**Pull request closed without merging**